src/Filesystem/File.php

```diff
@@ -441,11 +442,17 @@ public function root(): string|null
 	 * @return $this
 	 *
 	 * @throws \Kirby\Exception\InvalidArgumentException When the model does not use the `Kirby\Filesystem\IsFile` trait
+	 *                                                   or does not have `root` and `url` methods
 	 */
 	protected function setModel(object|null $model = null): static
 	{
-		if ($model !== null && in_array(IsFile::class, class_uses($model)) !== true) {
-			throw new InvalidArgumentException('The model object must use the "Kirby\Filesystem\IsFile" trait');
+		// TODO: create an interface instead of checking methods
+		if (
+			$model !== null &&
+			method_exists($model, 'root') !== true &&
+			method_exists($model, 'url') !== true
```

Comment:
This should be an or check because both methods must exist:

```suggestion
			(
				method_exists($model, 'root') !== true ||
				method_exists($model, 'url') !== true
			)
```

For cleaner code, we could move the `null` case to the top:

```php
if ($model === null) {
	$this->model = null;
	return $this;
}

if (
	method_exists($model, 'root') !== true ||
	method_exists($model, 'url') !== true
) {
	throw new InvalidArgumentException('The model object must use the "Kirby\Filesystem\IsFile" trait or have "root" and "url" methods');
}

$this->model = $model;
return $this;
```

Reply:
Thanks for suggestion 👍